pycbc/workflow/core.py

```diff
@@ -380,7 +409,8 @@ def add_ini_opts(self, cp, sec):
                 # stuff later.
                 self.unresolved_td_options[opt] = value
             else:
-                self.common_options += [opt, value]
+                # This option comes from the config file(s)
+                self.add_or_replace_common_options(opt, value)
```

Comment:
I don't understand the change here. I don't think I should be allowing to *replace* options here, and I don't know why a similar change for the other two cases above wouldn't also be needed? I think if we enforce that:
```
[coinc]
option1 = 10

[coinc-defaultvalues]
option1=15
```
is not allowed, then is this needed?

Reply:
This would be needed in the case of

```
[coinc-defaultvalues]
option1 = 10

[coinc-subsec]
option1 = 15
```

As the defaultvalues would be set first, and then the subsec would be overwriting it

Reply:
You're right there should be the same for the above examples though - that was missed

Reply:
This should be fine:
```
[coinc-defaultvalues]
option1 = 10

[coinc-subsec]
option1 = 15
```

This should not be:

```
[coinc]
option1 = 10

[coinc-defaultvalues]
option1=15
```

Reply:
But for the bigger question of does this need changing:

If this line needs a change, the other two cases (time-dependent options and File options) might need changes as well. I think the time-dependent option is fine, whereas the File option probably will not be.

Let's look closer at where this section is used, in particular it is *only* called from here:

https://github.com/gwastro/pycbc/blob/158e038a8e1186c686523c3da686aeaede5ad90b/pycbc/workflow/core.py#L571

(perhaps this should be _add_ini_opts to avoid anyone trying to call it from anywhere else, which I think would break things).

Now this will be called on all sections and subsections (not defaultvalues explicitly), and you then call `_defaultvalues` again for all of these .... This does open up the possibility of `[section-tag1-defaultvalues]` and `[section-tag2-defaultvalues]` which would conflict and we probably want to avoid.

You are also assuming that `[section]` is always sent to `add_ini_opts` before any subsections. It is, but if anyone changes that this will break in 'interesting' ways.

I would try and implement a model where you use a new dict defined here:

https://github.com/gwastro/pycbc/blob/158e038a8e1186c686523c3da686aeaede5ad90b/pycbc/workflow/core.py#L568

which would keep track of all options added (not values). Then I would explicitly call `add_ini_opts` after this for loop:

https://github.com/gwastro/pycbc/blob/158e038a8e1186c686523c3da686aeaede5ad90b/pycbc/workflow/core.py#L575

with `self.name + "-defaultvalues"`. This would also have a new keyword argument for "don't override existing values". The default behaviour would be that we fail if an option is already in the new dict.

Reply:
I'll look more closely later, but banksim and faithsim do not use *this* add_ini_opts function, they use a different, but same named, function in glue (and need to move over to pegasus).